HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

shamser opened the pull request:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

```
Set tautological-compare compiler flag to be an error so 'this==null' comparision is treated as errors
tautlogical-compare also treats &reference==null as errors also
Code to remove &reference==null check by either removing the check where it is unnecessary or converting the reference to a pointer
Remove unnecessary check of IPropertyTree & for nullness that was returned in IPropertyTreeIterator.
ForEach() calls IPropertyTreeIterator's isValid member function to ensure item is vaid.
IAtom: Remove multiple aliases to access getNamePtr() from IAtom
Add global str(IAtom *) function to safely access getNamePtr() even if IAtom * argument is null
IIdAtom: Remove multiple aliases to access getNamePtr() from IAtom
Add global str(IIdAtom *) function to safely access getNamePtr() even if IIdAtom * argument is null
Replace all usages of the aliases removed previously with a call to the newly create str() function
IIdAtom: remove the lower() member variable and replace with global lower(IIdAtom *p) function
Modify cmake commonSetup.cmake to have this==null treated as compilation errors by CLang compilers
Remove checks of reference variable’s address
secuser pointer checked for nullness before passing in argument as reference (working.cpp: CSecureConstWUIterator::checkScope)
user pointer checked for nullness before passing in argument as reference (ws_DfuService.cpp: getFilePermissions)
Change user to reference in getFilePermission() (ws_dfuService.cpp). The caller checks user for nullness, so appropriate to have caller pass user as a reference.
modify declaration of str() and lower() in jhash.cpp with jlib_decl (to expose externally in dlls)
```

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

Set tautological-compare compiler flag to be an error so 'this==null' comparision is treated as errors
tautlogical-compare also treats &reference==null as errors also
Code to remove &reference==null check by either removing the check where it is unnecessary or converting the reference to a pointer
Remove unnecessary check of IPropertyTree & for nullness that was returned in IPropertyTreeIterator.
ForEach() calls IPropertyTreeIterator's isValid member function to ensure item is vaid.
IAtom: Remove multiple aliases to access getNamePtr() from IAtom
Add global str(IAtom *) function to safely access getNamePtr() even if IAtom * argument is null
IIdAtom: Remove multiple aliases to access getNamePtr() from IAtom
Add global str(IIdAtom *) function to safely access getNamePtr() even if IIdAtom * argument is null
Replace all usages of the aliases removed previously with a call to the newly create str() function
IIdAtom: remove the lower() member variable and replace with global lower(IIdAtom *p) function
Modify cmake commonSetup.cmake to have this==null treated as compilation errors by CLang compilers
Remove checks of reference variable’s address
secuser pointer checked for nullness before passing in argument as reference (working.cpp: CSecureConstWUIterator::checkScope)
user pointer checked for nullness before passing in argument as reference (ws_DfuService.cpp: getFilePermissions)
Change user to reference in getFilePermission() (ws_dfuService.cpp). The caller checks user for nullness, so appropriate to have caller pass user as a reference.
modify declaration of str() and lower() in jhash.cpp with jlib_decl (so exposed correctly in dlls)

shamser committed:
HPCC-13496 'this' pointer cannot be null in well-defined C++ code

- Undo changes to archive.cpp

shamser commented:
Please see original code review comments here: https://github.com/hpcc-systems/HPCC-Platform/pull/7410.  @richardkchapman

richardkchapman commented on ecl/hql/hqlparse.cpp:
```diff
@@ -6,7 +6,7 @@
     You may obtain a copy of the License at
 
-       http://www.apache.org/licenses/LICENSE-2.0
+       http://www.apache.org/licenses/LICENSE-2.04
```
This change looks unrelated

shamser replied:
It is unrelated.  Fixed.

richardkchapman commented:
A general comment: you have squashed down most of the original commits to a single one - this is often a good idea but in a case like this where there are two distinct phases to the change (the references that were done manually/carefully/each on their own merits, and need careful review, versus the str() change that was done automatically and does not need such careful review), it's a lot easier to review (and indeed to refer back to a change in the history should we need to) if you keep them separate.

Not worth fixing now since the squash has been done, but something to bear in mind for the future.

I can merge as it is, though there are a couple of minor comments you may want to address first. It's going to need a manual merge as it stands, since there is a clashing change merged into master since you branched/last rebased. I can do that (the clash is trivial) but normally I ask people to rebase and repush in such cases.

richardkchapman commented on ecl/hql/hqlexpr.ipp:
```diff
@@ -399,7 +399,7 @@ protected:
 public:
     virtual IHqlExpression *clone(HqlExprArray &newkids);
-    virtual IAtom * queryName() const { return id->lower(); }
+    virtual IAtom * queryName() const { return id ? id->queryLower() : NULL; }
```
I'm curious why you have changed some instances of lower(id) to the construct "id ? id->queryLower() : NULL" and others to say "lower(id)"

Would be better to use lower(id) in all cases I think.

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

Undo change in version number

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

Modify as per code review comments: use the standard access function

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

shamser commented:
@richardkchapman Noted regarding keeping unrelated commits separate.  Code amended as per code review comments.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

ghalliday commented on system/jlib/jhash.hpp:
```diff
@@ -49,11 +49,8 @@ interface jlib_decl IAtom : extends IMapping
 {
  public:
     virtual const char * getNamePtr() const = 0;
```
can we rename this function to something that would be sensible to call.

e.g., queryStr()

Some of the instances that have been converted could call this directly if we know that this != NULL.

ghalliday commented on common/thorhelper/thorrparse.cpp:
```diff
@@ -2022,7 +2022,7 @@ void RegexNamedPattern::toXMLattr(StringBuffer & out, RegexXmlState & state)
     RegexPattern::toXMLattr(out, state);
     if (def->queryName())
-        out.appendf(" name=\"%s\"", def->queryName()->str());
+        out.appendf(" name=\"%s\"", str(def->queryName()));
```
could call def->queryName()->queryStr() since cannot be null

ghalliday commented on ecl/hql/hqlatoms.hpp:
```diff
@@ -448,7 +448,7 @@ extern HQL_API IAtom * xmlnsAtom;
 extern HQL_API IAtom * xpathAtom;
 
-inline bool isInternalAttributeName(IAtom * name) { return (name->str()[0] == '$'); }
+inline bool isInternalAttributeName(IAtom * name) { return (str(name)[0] == '$'); }
```
minor: could call ->queryStr() since it would crash if it was NULL

ghalliday commented on ecl/hql/hqlerror.cpp:
```diff
@@ -206,7 +206,7 @@ class HQL_API FileErrorReceiver : public ErrorReceiverSink
         StringBuffer msg;
         error->errorMessage(msg);
-        if (!filename) filename = isError(severity) ? "" : *unknownAtom;
+        if (!filename) filename = isError(severity) ? "" : str(unknownAtom);
```
unknownAtom->queryStr()

ghalliday replied:
I won't point any more out unless they are really critical.  Not really necessary to change any of them.

ghalliday commented on ecl/hql/hqlexpr.cpp:
```diff
@@ -7944,10 +7944,10 @@ void CHqlScope::throwRecursiveError(IIdAtom * searchName)
     StringBuffer filename;
     if (fullName)
         filename.append(fullName).append('.');
-    filename.append(*searchName);
```
This change looks suspicious.  It should be 
   filename.append(str(searchName));
or even
   filename.append(searchName)

Previously the operator * was defined so that *myatom was the same as myatom->str();

*str(myAtom) will return just the 1st character

ghalliday commented on ecl/hql/hqlexpr.cpp:
```diff
@@ -7944,10 +7944,10 @@ void CHqlScope::throwRecursiveError(IIdAtom * searchName)
 
     StringBuffer msg("Definition of ");
-    msg.append(*searchName).append(" contains a recursive dependency");
+    msg.append(*str(searchName)).append(" contains a recursive dependency");
```
same here

ghalliday commented on ecl/hql/hqlexpr.cpp:
```diff
@@ -8136,13 +8136,13 @@ IHqlExpression *CHqlRemoteScope::lookupSymbol(IIdAtom * searchName, unsigned loo
     if (fullName)
         filename.append(fullName).append('.');
-    filename.append(*searchName);
+    filename.append(*str(searchName));
```
and here, and several lines below.

ghalliday commented on ecl/hql/hqlexpr.ipp:
```diff
@@ -1630,7 +1630,7 @@ public:
     virtual IHqlExpression *clone(HqlExprArray &newkids);
     virtual bool equals(const IHqlExpression & other) const;
-    virtual IAtom * queryName() const { return unnamedId->lower(); }
+virtual IAtom * queryName() const { return unnamedId ? unnamedId->queryLower() : NULL; }
```
minor: indentation

ghalliday commented on esp/platform/sechandler.cpp:
```diff
@@ -95,6 +95,8 @@ bool SecHandler::authorizeSecFeature(const char * pszFeatureUrl, const char* Use
     if(pSecondaryUser.get()== NULL)
     {
         pSecondaryUser.setown(m_secmgr->createUser(UserID));
+	if (!pSecondaryUser.get())
```
minor: indentation (this line and next)
trivial: you do not need to call .get()

ghalliday commented on esp/services/ws_access/ws_accessService.cpp:
```diff
@@ -727,7 +727,8 @@ bool Cws_accessEx::onAddUser(IEspContext &context, IEspAddUserRequest &req, IEsp
         try
         {
-            secmgr->addUser(*user.get());
+	    if (user.get())
```
minor: indentation

ghalliday commented:
@shamser good work.  I've had an initial scan through and picked out a few issues - only one serious.
I'll have another more careful review once those are addressed.

ghalliday commented on ecl/hql/hqlgram.y:
```diff
@@ -3931,22 +3931,22 @@ attrib
                         {
                             parser->reportWarning(CategoryDeprecated, SeverityError, WRN_OBSOLETED_SYNTAX,$1.pos,"Syntax obsoleted; use alternative: id = '<string constant>'");
-                            $$.setExpr(createAttribute($1.getId()->lower(), createConstant(*$3.getId())));
+                            $$.setExpr(createAttribute(lower($1.getId()), createConstant(*str($3.getId()))));
```
createConstant(str(...)) no *

richardkchapman commented:
Looks like you changed once instance of the "id ? id->queryLower() : NULL; " paradigm - but there were several more.

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

- getNamePtr() in IIdAtom and IAtom class renamed to queryStr()
- directly call queryStr() member function  of IIdAtom and IAtom
instead of using str when object known to be non-null
- fix where *str() used when str() should have been fixed
- use lower(id) instead of "id ? id->queryLower() : NULL; "
- Replace tabs with spaces

shamser commented:
@richardkchapman @ghalliday Thanks for your comments.  Changes as per code review comments complete.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

ghalliday commented on ecl/hql/hqlatoms.hpp:
```diff
@@ -448,7 +448,7 @@ extern HQL_API IAtom * xmlnsAtom;
 extern HQL_API IAtom * xpathAtom;
 
-inline bool isInternalAttributeName(IAtom * name) { return (name->str()[0] == '$'); }
+inline bool isInternalAttributeName(IAtom * name) { return name ? false : (name->queryStr()[0] == '$'); }
```
trivial: name cannot be NULL, so code can be simplified

ghalliday replied:
(test is also the wrong way round)

ghalliday commented on ecl/hql/hqlerror.cpp:
```diff
@@ -206,7 +206,7 @@ class HQL_API FileErrorReceiver : public ErrorReceiverSink
         StringBuffer msg;
         error->errorMessage(msg);
-        if (!filename) filename = isError(severity) ? "" : *unknownAtom;
+        if (!filename) filename = isError(severity) || unknownAtom ? "" : unknownAtom->queryStr();
```
The test for unknownAtom is unnecessary, and the wrong way around.

ghalliday commented:
@shamser I checked more carefully this time.  Just two issues picked up.

There will also need to be a 5.4 pull request which just adds the str() and lower() functions to avoid future merge errors.

We are also likely to need a pull request to the legacy repository (ln) to fix any instances there.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented:
I'm getting compilation errors with this branch:

/Users/rchapman/git/HPCC-Platform/plugins/mysql/mysqlembed.cpp:321:51: error: no member named 'str' in 'IAtom'
        msg.appendf(" for field %s", field->name->str());
/Users/rchapman/git/HPCC-Platform/plugins/mysql/mysqlembed.cpp:607:60: error: no member named 'str' in 'IAtom'
            failx("Error fetching column %s", field->name->str());

I'm guessing you compiled on a system with mysql not installed, so the optional MySQL plugin will have been skipped. You should check for similar issues in the other embedded language plugins.

shamser committed:
HPCC-13496 Fix compilation Error: 'this' pointer and references cannot be null

Modify callers to call the IAtom::queryStr() rather than IAtom::str()

shamser commented:
@richardkchapman plugins checked to make sure correct member function called.  Just found the 2 you discovered.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented:
Merged manually to address required rebase
